File: Kotlin/ByTier/G/2137_G_Cry_Me_a_River.kt
// Problem: CF 2137 G - Cry Me a River
// https://codeforces.com/contest/2137/problem/G

/*
 * Problem: G. Cry Me a River
 *
 * Purpose:
 *   This code solves a game played on a directed acyclic graph (DAG) between two players, Cry and River.
 *   Each node starts blue; when queried with type 1, a node is painted red.
 *   Player Cry moves first, and the game ends when either:
 *     - The token reaches a node with no outgoing edges (Cry wins).
 *     - The token reaches a red node (River wins).
 *   The queries of type 2 ask whether Cry can win from a given starting node with optimal play.
 *
 * Algorithm:
 *   - Use topological sorting to process nodes from leaves up to root.
 *   - At each node, compute "A" and "B" properties:
 *     - A[v]: Player Cry wins from node v (i.e., there is a path for Cry to a leaf).
 *     - B[v]: All paths from v reach only red or leaf nodes (i.e., River cannot win from here if he plays optimally).
 *   - Maintain counters for non-red children, and red children during updates.
 *   - BFS/DFS on reverse graph to propagate changes when updating node color to red.
 *
 * Time Complexity:
 *   O(n + m + q) per test case in total amortized time due to efficient updates and propagation.
 * Space Complexity:
 *   O(n + m) for storing the graph and auxiliary arrays.
 */

import java.io.BufferedInputStream
import java.lang.StringBuilder
import java.util.ArrayDeque

private class FastScanner {
    private val input = BufferedInputStream(System.`in`)
    private val buffer = ByteArray(1 shl 16)
    private var len = 0
    private var ptr = 0
    private fun readByte(): Int {
        if (ptr >= len) {
            len = input.read(buffer)
            ptr = 0
            if (len <= 0) return -1
        }
        return buffer[ptr++].toInt()
    }
    fun nextInt(): Int {
        var c = readByte()
        while (c <= 32 && c >= 0) c = readByte()
        var sign = 1
        if (c == '-'.code) { sign = -1; c = readByte() }
        var res = 0
        while (c > 32) {
            res = res * 10 + (c - '0'.code)
            c = readByte()
        }
        return res * sign
    }
}

fun main() {
    val fs = FastScanner()
    val t = fs.nextInt()
    val out = StringBuilder()
    repeat(t) {
        val n = fs.nextInt()
        val m = fs.nextInt()
        val q = fs.nextInt()
        val from = IntArray(m)
        val to = IntArray(m)
        val outDeg = IntArray(n)
        val inDegRev = IntArray(n)
        for (i in 0 until m) {
            val u = fs.nextInt() - 1
            val v = fs.nextInt() - 1
            from[i] = u
            to[i] = v
            outDeg[u]++
            inDegRev[v]++
        }
        val adj = Array(n) { IntArray(0) }
        val rev = Array(n) { IntArray(0) }
        for (i in 0 until n) adj[i] = IntArray(outDeg[i])
        for (i in 0 until n) rev[i] = IntArray(inDegRev[i])
        val posOut = IntArray(n)
        val posIn = IntArray(n)
        for (i in 0 until m) {
            val u = from[i]
            val v = to[i]
            adj[u][posOut[u]++] = v
            rev[v][posIn[v]++] = u
        }
        val indeg = IntArray(n)
        for (i in 0 until m) indeg[to[i]]++
        val dqTopo = ArrayDeque<Int>()
        for (i in 0 until n) if (indeg[i] == 0) dqTopo.add(i)
        val topo = IntArray(n)
        var ti = 0
        while (dqTopo.isNotEmpty()) {
            val v = dqTopo.removeFirst()
            topo[ti++] = v
            for (w in adj[v]) {
                indeg[w]--
                if (indeg[w] == 0) dqTopo.add(w)
            }
        }
        val red = BooleanArray(n) 
        val A = BooleanArray(n)   
        val B = BooleanArray(n)   
        val countNonRed = IntArray(n) 
        val numRedChildren = IntArray(n) 
        val countNonRedTrueA = IntArray(n) 
        val numGood = IntArray(n) 
        for (i in 0 until n) {
            countNonRed[i] = outDeg[i]
            numRedChildren[i] = 0
        }
        for (idx in n - 1 downTo 0) {
            val v = topo[idx]
            if (outDeg[v] == 0) {
                A[v] = true
                B[v] = true
                countNonRedTrueA[v] = 0
                numGood[v] = 0
            } else {
                var ca = 0
                var gb = 0
                for (w in adj[v]) {
                    if (A[w]) ca++
                    if (B[w] && !red[w]) gb++
                }
                countNonRedTrueA[v] = ca
                numGood[v] = gb
                A[v] = (numGood[v] > 0)
                B[v] = (numRedChildren[v] == 0 && countNonRedTrueA[v] == countNonRed[v])
            }
        }
        val queue = ArrayDeque<Int>()
        for (qi in 0 until q) {
            val type = fs.nextInt()
            val u = fs.nextInt() - 1
            if (type == 1) {
                if (red[u]) {
                } else {
                    val oldA = A[u]
                    val oldB = B[u]
                    red[u] = true
                    A[u] = false
                    B[u] = false
                    for (p in rev[u]) {
                        countNonRed[p]--
                        numRedChildren[p]++
                        if (oldA) countNonRedTrueA[p]--
                        if (oldB) numGood[p]--
                        queue.add(p)
                    }
                    while (queue.isNotEmpty()) {
                        val v = queue.removeFirst()
                        val prevA = A[v]
                        val prevB = B[v]
                        val newA: Boolean
                        val newB: Boolean
                        if (red[v]) {
                            newA = false
                            newB = false
                        } else if (outDeg[v] == 0) {
                            newA = true
                            newB = true
                        } else {
                            newA = (numGood[v] > 0)
                            newB = (numRedChildren[v] == 0 && countNonRedTrueA[v] == countNonRed[v])
                        }
                        if (newA == prevA && newB == prevB) continue
                        val deltaA = (if (newA) 1 else 0) - (if (prevA) 1 else 0)
                        val deltaB = (if (newB) 1 else 0) - (if (prevB) 1 else 0)
                        A[v] = newA
                        B[v] = newB
                        if (deltaA != 0 || deltaB != 0) {
                            for (p in rev[v]) {
                                if (deltaA != 0) countNonRedTrueA[p] += deltaA
                                if (deltaB != 0) numGood[p] += deltaB
                                queue.add(p)
                            }
                        }
                    }
                }
            } else {
                out.append(if (A[u]) "YES\n" else "NO\n")
            }
        }
    }
    print(out.toString())
}


// https://github.com/VaHiX/codeForces/

File: CPP/ByRound/2137/2137_G_Cry_Me_a_River.cpp
// Problem: CF 2137 G - Cry Me a River
// https://codeforces.com/contest/2137/problem/G

/*
G. Cry Me a River
Algorithms/Techniques: Topological sorting, BFS-based game simulation with memoization on states.
Time Complexity: O(n + m + q)
Space Complexity: O(n + m)

This code handles a directed acyclic graph (DAG) problem involving a two-player game between Cry and River.
Each node is initially blue. For queries of type 1, we color a node red.
For queries of type 2, we determine if Cry wins when the game starts from a given node under optimal play.

The solution uses:
- Reverse topological order processing via BFS to compute winning conditions.
- State tracking using a 2D array f[i][0] and f[i][1], representing whether a node leads to win for Cry (f[i][0]) or River (f[i][1]).
- Special handling of game states where players can immediately win or lose based on node out-degree and color.
*/

#include <algorithm>
#include <iostream>
#include <queue>

using namespace std;
const int N = 4e5 + 5;
int head[N], num, n, m, k, in[N], deg[N];
struct edge {
  int to, next;
} g[N * 2];
void addd(int u, int v) {
  num++;
  g[num].to = v;
  g[num].next = head[u];
  head[u] = num;
}
bool col[N], ans[N], vis[N], f[N][2];
struct node {
  int opt, x;
} a[N];
queue<int> q;
void work(int x) {
  col[x] = 1;
  if (f[x][0])
    q.push(x);
  if (f[x][1])
    q.push(x + n); // Pushing x+n to distinguish in BFS
  while (!q.empty()) {
    int u = q.front();
    q.pop();
    if (u > n) { // Handling nodes for "River wins" condition
      u -= n;
      for (int i = head[u]; i; i = g[i].next) {
        int v = g[i].to;
        if (f[v][0] == 0) { // If Cry can win from this node yet not set
          f[v][0] = 1;
          if (col[v])
            q.push(v); // If this node is colored, push it to queue for further propagation
        }
      }
    } else { // Handling nodes for "Cry wins" condition
      for (int i = head[u]; i; i = g[i].next) {
        int v = g[i].to;
        in[v]--;
        if (in[v] == 0) {
          f[v][1] = 1; // Marking that River can win from 'v'
          if (col[v])
            q.push(v + n); // If colored, push related to river queue
        }
      }
    }
  }
}
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);
  int T;
  cin >> T;
  while (T--) {
    int q;
    cin >> n >> m >> q;
    int u, v;
    num = 0;
    for (int i = 1; i <= n; i++)
      head[i] = f[i][0] = f[i][1] = in[i] = deg[i] = col[i] = 0;
    for (int i = 1; i <= m; i++) {
      cin >> u >> v;
      addd(v, u); // Edge from v to u (reverse direction)
      in[u]++;
      deg[u]++;
    }
    for (int i = 1; i <= n; i++)
      if (!in[i]) // Nodes with no incoming edges are terminal
        f[i][0] = f[i][1] = 1;
    for (int i = 1; i <= q; i++) {
      cin >> a[i].opt >> a[i].x;
      if (a[i].opt == 1)
        col[a[i].x] = 1; // Mark node as red (colored)
    }
    for (int i = 1; i <= n; i++) {
      if (!col[i])
        q++, a[q].opt = 1, a[q].x = i; // Add uncolored nodes to query list
      col[i] = 0;
    }
    for (int i = q; i >= 1; i--) {
      if (a[i].opt == 1) {
        work(a[i].x); // Process coloring updates
      } else {
        ans[i] = (col[a[i].x] & f[a[i].x][0]); // Query result based on current state
      }
    }
    for (int i = 1; i <= q; i++)
      if (a[i].opt == 2)
        cout << (ans[i] ? "YES\n" : "NO\n"); // Output results of type 2 queries
  }
}


// https://github.com/VaHiX/codeForces/